### PR TITLE
Capture the Cassandra 4.4 logical server in server.address

### PR DESCRIPTION
**Pull request by trask**
This PR is part of #19899 and applies the configured database target rules in [open-telemetry/semantic-conventions#4058](https://github.com/open-telemetry/semantic-conventions/pull/4058) to Cassandra 4.4. Stable spans report configured contact points in `server.address` rather than the selected coordinator; legacy coordinator and proxy telemetry are preserved, with operation peer information handled separately by #19883.

Library users supply the complete original contact-point list, including both builder and driver-configuration sources:

```java
CqlSession tracedSession = cassandraTelemetry.wrap(session, contactPoints);
```

The existing `wrap(session)` still traces operations but omits stable `server.address` and `server.port`. This follows the MongoDB library pattern in #19839 rather than trying to reconstruct original configuration from resolved driver metadata.

The javaagent continues to combine captured builder contact points with `basic.contact-points` automatically. Targets are captured once when wrapping the session, so later topology and coordinator changes do not affect them.

### Multi-endpoint handling

- A single endpoint omits its port from `server.address`, and a non-default port is recorded in `server.port`.
- Multiple endpoints omit ports when all use the default `9042` port.
- When any of multiple endpoints uses a non-default port, every endpoint includes its port in `server.address` and `server.port` is omitted.
- Endpoints are normalized and sorted, configured duplicates are preserved, and at most the first five resulting endpoints are recorded.

### Omission rules

Unsafe or malformed values, implicit targets, and custom or SNI discovery endpoints omit the stable target instead of falling back to the coordinator.

Builds on #19872.

### CODE REVIEW

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraSqlAttributesGetter.java [hunk 1]
```diff
@@ -75,4 +91,17 @@ public InetSocketAddress getNetworkPeerInetSocketAddress(
...
+    if (!emitStableDatabaseSemconv()) {
+      return null;
+    }
+    for (Node node : request.getSession().getMetadata().getNodes().values()) {
+      if (node.getEndPoint() instanceof SniEndPoint) {
+        return null;
+      }
+    }
+    return request.getServerTarget();
+  }
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraSqlAttributesGetter.java]: Addressed in f9caea20f59e7e7ea799af0969a8272c4879a8e5.

Analysis: CassandraServerTarget already rejects SNI endpoints when the session is wrapped. The request stores that result for its lifetime. Reading the stored target directly removes the metadata scan and makes start-time emission use the same request predicate as end-time preservation. The SNI endpoint tests now model the production invariant by passing no configured target.

Upsides: Span start no longer scans cluster metadata, and start and end agree on whether the configured target was emitted.

Downsides: The getter relies on CassandraServerTarget construction to keep rejecting SNI sessions. No material runtime downside identified.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraAttributesExtractor.java [hunk 2]
```diff
@@ -155,28 +156,66 @@ public void onEnd(
...
+      } else {
+        // Legacy semconv keeps recording the proxy as the server.
+        updateLegacySniServerAddressAndPort(attributes, sniEndPoint);
+      }
+      return;
+    }
+    // Preserve the configured target that stable semconv records on span start.
+    if (emitStableDatabaseSemconv() && request.getServerTarget() != null) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraAttributesExtractor.java]: Addressed in f9caea20f59e7e7ea799af0969a8272c4879a8e5.

Analysis: CassandraServerTarget already rejects SNI endpoints when the session is wrapped. The request stores that result for its lifetime. Reading the stored target directly removes the metadata scan and makes start-time emission use the same request predicate as end-time preservation. The SNI endpoint tests now model the production invariant by passing no configured target.

Upsides: Span start no longer scans cluster metadata, and start and end agree on whether the configured target was emitted.

Downsides: The getter relies on CassandraServerTarget construction to keep rejecting SNI sessions. No material runtime downside identified.

File: instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/SessionBuilderInstrumentation.java [hunk 3]
```diff
@@ -46,8 +51,11 @@ public static class BuildAdvice {
      */
     @AssignReturned.ToReturned
     @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
-    public static CompletionStage<?> injectTracingSession(@Advice.Return CompletionStage<?> stage) {
-      return stage.thenApply(new CompletionStageFunction());
+    public static CompletionStage<?> injectTracingSession(
+        @Advice.Return CompletionStage<?> stage,
+        @Advice.FieldValue("programmaticContactPoints") Set<EndPoint> programmaticContactPoints) {
+      return stage.thenApply(
+          new CompletionStageFunction(telemetry(), new HashSet<>(programmaticContactPoints)));
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/SessionBuilderInstrumentation.java:57]: No code change.

Analysis: The supported DataStax 4.4 and Apache 4.18 drivers initialize `programmaticContactPoints` to a new `HashSet` in the `SessionBuilder` constructor. The Apache driver also resets it to a new `HashSet` when it clears contact points. The driver does not use null to represent an unset value.

Upsides: Keeping the direct copy matches the driver's non-null field invariant and avoids an unreachable fallback.

Downsides: A custom SessionBuilder subclass that breaks the driver's protected-field invariant would still skip wrapping under the advice's existing Throwable suppression.

File: instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java [hunk 4]
```diff
@@ -9,15 +9,15 @@
 import io.opentelemetry.instrumentation.api.incubator.config.internal.DbConfig;
 import io.opentelemetry.instrumentation.cassandra.v4_4.CassandraTelemetry;
 
-final class CassandraSingletons {
+public final class CassandraSingletons {
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java:12]: Addressed in 11b8380fea749e5bec6fe1aea298b984ef029404.

Analysis: Byte Buddy copies the advice body into SessionBuilder, so the transformed Datastax class directly accesses CassandraSingletons. Package-private visibility causes IllegalAccessError before the helper constructor runs. CassandraTelemetry is already passed into CompletionStageFunction through its constructor, but that does not remove the transformed class's access to the singleton. Public visibility is required for this javaagent helper boundary and does not create a supported application API.

Upsides: SessionBuilder can run the injected advice and wrap sessions without IllegalAccessError.

Downsides: The javaagent implementation class remains publicly accessible at the bytecode level, though it is not a supported API.

File: instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java [hunk 5]
```diff
@@ -9,15 +9,15 @@
...
   private static final CassandraTelemetry telemetry =
       CassandraTelemetry.builder(GlobalOpenTelemetry.get())
           .setQuerySanitizationEnabled(
               DbConfig.isQuerySanitizationEnabled(GlobalOpenTelemetry.get(), "cassandra"))
           .build();
 
-  static CassandraTelemetry telemetry() {
+  public static CassandraTelemetry telemetry() {
     return telemetry;
   }
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java:22]: Addressed in 11b8380fea749e5bec6fe1aea298b984ef029404.

Analysis: Byte Buddy copies the advice body into SessionBuilder, so the transformed Datastax class directly accesses CassandraSingletons. Package-private visibility causes IllegalAccessError before the helper constructor runs. CassandraTelemetry is already passed into CompletionStageFunction through its constructor, but that does not remove the transformed class's access to the singleton. Public visibility is required for this javaagent helper boundary and does not create a supported application API.

Upsides: SessionBuilder can run the injected advice and wrap sessions without IllegalAccessError.

Downsides: The javaagent implementation class remains publicly accessible at the bytecode level, though it is not a supported API.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      return uniqueContactPoints.values().iterator().next();
+    }
+    StringBuilder group = new StringBuilder();
+    for (String contactPoint : uniqueContactPoints.keySet()) {
+      if (group.length() > 0) {
+        group.append(',');
+      }
+      group.append(contactPoint);
+    }
+    return new CassandraServerTarget(group.toString(), null);
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java]: No code change.

Analysis: Multiple targets keep each target's port inside the comma-separated `server.address` value. Only the separate `server.port` attribute is omitted. The pull request description now states this distinction.

Upsides: The description matches the implemented attribute values and removes the ambiguity.

Downsides: No material downside identified.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    return target.build();
+  }
+
+  @Nullable
+  static DbServerTarget ofAddresses(Collection<InetSocketAddress> contactPoints) {
+    DbServerTargetBuilder target = DbServerTarget.builder(DEFAULT_PORT).setSorted(true);
+    for (InetSocketAddress contactPoint : contactPoints) {
+      target.addEndpoint(contactPoint);
+    }
+    return target.build();
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java]: Addressed in 6e2db379251dc043f7618d7dcceb0355e87b1b15.

Analysis: The explicit-address overload let invalid entries throw before the target builder could omit the target. Catching RuntimeException around endpoint collection and target construction applies the same omission behavior as the session overloads.

Upsides: Unsafe or malformed explicit addresses now omit server.address instead of escaping into session construction.

Downsides: The overload does not expose the specific malformed endpoint that caused omission.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+      host = host.substring(1, host.length() - 1);
+    }
+    try {
+      int port = Integer.parseInt(contactPoint.substring(separator + 1));
+      target.addEndpoint(host, port);
+      return new ContactPoint(host, port);
+    } catch (NumberFormatException ignored) {
+      target.addEndpoint((String) null, -1);
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java:110]: Addressed in 1836202b229d0b1d5d098ccc84371faaf32113fa.

Analysis: The last colon cannot distinguish an IPv6 address segment from a port. Requiring brackets whenever the parsed host contains a colon makes the configured contact point format unambiguous and keeps bare IPv6 literals from producing a false stable target.

Upsides: IPv6 contact points now produce a stable target only from the deterministic `[address]:port` form, and tests cover apparent-port edge cases.

Downsides: Previously accepted unbracketed IPv6 contact points with a port now omit the stable target.

File: instrumentation/cassandra/cassandra-common-4.0/testing/src/main/java/io/opentelemetry/cassandra/common/v4_0/AbstractCassandraTest.java [hunk 9]
```diff
@@ -192,6 +194,76 @@ void simpleStatementWithValues() {
...
+                            .hasNoParent()
+                            .hasAttributesSatisfyingExactly(
+                                satisfies(
+                                    NETWORK_TYPE,
+                                    emitStableDatabaseSemconv()
+                                        ? val -> val.isNull()
+                                        : val -> val.isIn("ipv4", "ipv6")),
+                                equalTo(
+                                    SERVER_ADDRESS,
+                                    emitStableDatabaseSemconv() ? configuredTarget : cassandraIp),
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-common-4.0/testing/src/main/java/io/opentelemetry/cassandra/common/v4_0/AbstractCassandraTest.java:238]: No code change.

Analysis: The first endpoint uses `cassandraPort`, which Testcontainers assigns as the mapped host port for container port 9042. The second endpoint explicitly uses 9042. These are mixed ports, so the target builder must inline both ports.

Upsides: Keeping the expectation preserves integration coverage for the mixed-port target format.

Downsides: No material downside identified.

File: instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java:12]: No code change.

Analysis: Byte Buddy makes the transformed Datastax `SessionBuilder` access `CassandraSingletons` from another package. A prior package-private version failed with `IllegalAccessError` before `CompletionStageFunction` could receive the telemetry instance. The class and accessor therefore need public bytecode visibility across the advice boundary.

Upsides: The injected session-builder advice can obtain telemetry and wrap completed sessions without a linkage error.

Downsides: The javaagent implementation class remains publicly accessible at the bytecode level, though it is not a supported application API.

File: instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/CassandraSingletons.java:22]: No code change.

Analysis: Byte Buddy makes the transformed Datastax `SessionBuilder` access `CassandraSingletons` from another package. A prior package-private version failed with `IllegalAccessError` before `CompletionStageFunction` could receive the telemetry instance. The class and accessor therefore need public bytecode visibility across the advice boundary.

Upsides: The injected session-builder advice can obtain telemetry and wrap completed sessions without a linkage error.

Downsides: The javaagent implementation class remains publicly accessible at the bytecode level, though it is not a supported application API.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+        continue;
+      }
+      if (configuredTarget.host.equals(target.host)) {
+        return true;
+      }
+      InetAddress configuredAddress = numericAddress(configuredTarget.host);
+      InetAddress targetAddress = numericAddress(target.host);
+      if (configuredAddress != null && configuredAddress.equals(targetAddress)) {
+        return true;
+      }
```
Comment (laurit) [inline instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java]: I guess this is to figure out whether 2 addresses are really the same, does it matter?

**Review by laurit — APPROVED** [review]

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraTelemetry.java [hunk 11]
```diff
@@ -42,4 +52,40 @@ public static CassandraTelemetryBuilder builder(OpenTelemetry openTelemetry) {
...
+   * com.datastax.oss.driver.api.core.config.DriverConfigLoader}, use {@link #wrap(CqlSession)}
+   * instead.
+   *
+   * <p>The contact points are captured when the session is wrapped and are used only to derive
+   * stable database server attributes. They do not change the session's connections or
+   * configuration.
+   *
+   * <p>{@code contactPoints} must contain every contact point supplied to the builder. Do not pass
+   * the current coordinator, discovered cluster nodes, or only a subset of the configured contact
+   * points.
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraTelemetry.java]: Clarified in [a41b46daa1](https://github.com/open-telemetry/opentelemetry-java-instrumentation/commit/a41b46daa1c0a7b631334fe41fed1103b41c7a9e). The overload now requires the collection to contain every original contact point from every configuration source; the standard overload is documented for sessions configured exclusively through driver configuration.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+      MetadataManager metadataManager = ((InternalDriverContext) context).getMetadataManager();
+      if (metadataManager.wasImplicitContactPoint()) {
+        return null;
+      }
+      DriverExecutionProfile config = context.getConfig().getDefaultProfile();
+      // The basic.contact-points option has no default, so the single-argument lookup would throw
+      // when a session names its contact points on the builder alone.
+      List<String> configuredContactPoints = config.getStringList(CONTACT_POINTS, emptyList());
+      DbServerTargetBuilder target = DbServerTarget.builder(DEFAULT_PORT);
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java]: Fixed in [b8694d25f3](https://github.com/open-telemetry/opentelemetry-java-instrumentation/commit/b8694d25f35749cdacd2d0688c7bb2bc86ec7057). All Cassandra target reconstruction paths now sort the complete rendered endpoint list before capping, including mixed configured and programmatic sources.

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+      target.addEndpoint((String) null, -1);
+      return;
+    }
+    String host = contactPoint.substring(0, separator);
+    if (host.startsWith("[")) {
+      if (!host.endsWith("]")) {
+        target.addEndpoint((String) null, -1);
+        return;
+      }
+      host = host.substring(1, host.length() - 1);
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java:100]: Addressed in 7ec3c533582f5a0785265470782b63a52404eec7.

Analysis: CassandraServerTarget removes brackets before passing the host to DbServerTargetBuilder. The builder therefore cannot tell that a DNS hostname was bracketed. Rejecting a bracketed host without a colon preserves valid IPv6 literals and drops malformed bracketed hostnames.

Upsides: Malformed bracketed hostnames now omit the complete server target, and the parameterized parser test covers this case.

Downsides: No material downside identified

File: instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java (hunk 10 above)
Comment (trask) [reply instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraServerTarget.java]: yeah, this was probably too much, simplified it, and require passing in the seeds into the library instrumentation similar to mongo (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/19839)
